lib/structuraid_core/engineering/optimization/rebar_constant_long.rb

```diff
@@ -0,0 +1,126 @@
+                    :current_rebar,
+                    :amount_of_rebars,
+                    :steps_log,
+                    :iteration_rebar,
+                    :iteration_rebars_amount,
+                    :iteration_rebar_mass
+
+        def run_optimization
+          @material = StructuraidCore::Materials::Steel.new(yield_stress: 420)
+          @current_mass = Float::INFINITY
+          @current_rebar = nil
+          @amount_of_rebars = nil
+
+          find_optimal_rebar
+        end
+
+        def find_optimal_rebar
+          OPTIONAL_REBAR_NUMBERS.each do |number|
+            iteration_rebars_and_amount(number)
+            break if iteration_rebars_amount < MINIMUM_REBAR_AMOUNT
+
+            rebar_spacing = space_to_cover / (iteration_rebars_amount - 1)
+            break if rebar_spacing > maximum_rebar_spacing
+
+            @iteration_rebar_mass = reinforcement_mass(iteration_rebar, iteration_rebars_amount)
+            next unless iteration_rebar_mass < current_mass
+
+            commit_results
+          end
+        end
+
+        def iteration_rebars_and_amount(number)
+          @iteration_rebar = StructuraidCore::Elements::Reinforcement::Rebar.new(number:, material:)
+          @iteration_rebars_amount = required_amaunt_of_rebar(iteration_rebar)
+        end
+
+        def commit_results
+          @current_mass = iteration_rebar_mass
+          @current_rebar = iteration_rebar
+          @amount_of_rebars = iteration_rebars_amount
+          @steps_log << "rebar #{@current_rebar.number}: #{iteration_rebars_amount} ----> #{iteration_rebar_mass}"
+          puts "rebar #{@current_rebar.number}: #{iteration_rebars_amount} ----> #{iteration_rebar_mass}"
+        end
+
+        def required_amaunt_of_rebar(rebar)
+          calculated = required_reinforcement_area / rebar.area
+          round_amount(calculated.round)
+        end
+
+        def amount_of_rebars_based_on_spacing(space_to_cover, spacing)
+          calculated = space_to_cover / spacing
+          round_amount(calculated.round)
+        end
+
+        def round_amount(calculated_amount)
+          integer_amount = calculated_amount.round
+          return integer_amount + 1 if integer_amount < calculated_amount
+
+          integer_amount
+        end
+
+        def reinforcement_mass(rebar, amount_of_rebars)
+          rebar.mass * amount_of_rebars
+        end
+
+        def successful_result
+          { rebar: current_rebar, amount_of_rebars:, message: SUCCESSFUL_RESULT_MESSAGE, result_code: 0 }
+        end
+
+        def unsuccessful_result
+          {
+            rebar: StructuraidCore::Elements::Reinforcement::Rebar.new(
+              number: OPTIONAL_REBAR_NUMBERS.first,
+              material:
+            ),
+            amount_of_rebars: amount_of_rebars_based_on_spacing(space_to_cover, maximum_rebar_spacing),
+            message: UNSUCCESSFUL_RESULT_MESSAGE,
+            result_code: 1
+          }
+        end
```

Comment:
What if, instead of returning a hash as a result we create a struct here inside the class and return an instance of it on each scenario (successful vs. unsuccessful), that way we ensure we are returning the same interface for both states and keep some consistency.

```ruby
Result = Struct.new(:rebar, :amount_of_rebars, :message, :result_code)
```

I have also some concerns about `message` and `result_code`:
1. What if we just use `result_code` as a symbol with a few explicit possible values, like:
 - `result_code: :success` when the optimization is successful
 - `result_code: :maximum_spacing_unsatisfied`
2. It looks like `message` is some output you want to send back to the user, right? I think it would add a lot of complexity to the class. If we start adding messages like that everywhere all over our code we will get crazy when we need to refactor them to refine the UI. Think for example of internationalization or stuff like that, it's a whole different responsibility. If we just send a result_code we can delegate that responsibility of determining the message to the class that calls this one, or somewhere else just depending on the result_code that this one is returning

I think the only messages we have left so far like that have been on the designer organizer, and I think we can agree to something like leaving those messages only there because so far it looks like those organizers are the most user-facing calls we are going to have on this gem: everything we build on this gem is because at some point the designers are going to call them. We should not let them propagate across any other namespace

On the other hand, if you are thinking of these messages to be used only for debugging or logging but not to send back to the client, I'm fine with leaving them here because that kind of thing will hopefully remain pretty simple over time